### PR TITLE
[PROTOCOL-288] Aave dapp

### PR DESCRIPTION
**Pull request by sypeer**
It is:
- Initial Aave interfaces and dapp

### CODE REVIEW

File: contracts/base/escrow/dapps/Aave.sol [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+        );
+    }
+
+    /**
+        @notice This function withdraws the user's aTokens for a specific amount 
+        @param tokenAddress address of the token
+        @param amount amount of the underlying tokens to withdraw
+     */
+    function withdraw(address tokenAddress, uint256 amount) public onlyOwner() {
+        IAToken aToken = _getAToken(tokenAddress);
```
Comment (salazarguille) [inline contracts/base/escrow/dapps/Aave.sol:74]: Should we validate 0x0?

File: contracts/base/escrow/dapps/Aave.sol [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+    /**
+        @notice Grabs the aToken instance from the lending pool
+        @param tokenAddress The underlying asset address to get the aToken for
+        @return IAToken instance
+     */
+    function _getAToken(address tokenAddress) internal view returns (IAToken) {
+        return
+            IAToken(
+                _getAavePoolAddress(tokenAddress)
+                    .getReserveData(tokenAddress)
```
Comment (salazarguille) [inline contracts/base/escrow/dapps/Aave.sol:123]: Is it correct to use tokenAddress in both calls?

**Review by salazarguille — CHANGES_REQUESTED** [review]

**Comment by salazarguille** [comment]
Please verify the integration tests.

File: contracts/base/escrow/dapps/Aave.sol (hunk 1 above)
Comment (sypeer) [reply contracts/base/escrow/dapps/Aave.sol:74]: Doing this validation inside of the AssetSettings contract when it's being created

File: contracts/base/escrow/dapps/Aave.sol (hunk 2 above)
Comment (sypeer) [reply contracts/base/escrow/dapps/Aave.sol:123]: Yes: 

The first _getAavePoolAddress,  grabs the lending pool instance from the Aave lending pool address provider. 
https://github.com/teller-protocol/teller-protocol-v1/pull/252/files#diff-febb5fbb878382a74affe256f1bea611efac861dccff043b062b658463ae1a53R99

The second getReserveData gets the reserve data for the retrieved lending pool from the loaded interface

File: contracts/base/escrow/dapps/Aave.sol [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+    function withdraw(address tokenAddress, uint256 amount)
+        public
+        onlyBorrower
+    {
+        IAToken aToken = _getAToken(tokenAddress);
+        IAaveLendingPool aaveLendingPool = _getAavePoolAddress(tokenAddress);
+        uint256 aTokenBalanceBeforeWithdraw = aToken.balanceOf(address(this));
+        require(aTokenBalanceBeforeWithdraw >= 0, "NO_BALANCE_TO_WITHDRAW");
+        uint256 result =
+            aaveLendingPool.withdraw(tokenAddress, amount, address(this));
```
Comment (passabilities) [inline contracts/base/escrow/dapps/Aave.sol:86]: Is there a value we should be checking for?

File: contracts/base/escrow/dapps/Aave.sol [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+
+    /**
+        @notice Grabs the Aave lending pool instance from the Aave lending pool address provider
+        @param tokenAddress The token address to get the lending pool for
+        @return IAaveLendingPool instance address
+     */
+    function _getAavePoolAddress(address tokenAddress)
+        internal
+        view
+        returns (IAaveLendingPool)
```
Comment (passabilities) [inline contracts/base/escrow/dapps/Aave.sol:114]: Function is returning the LP interface, not just the address. Lets change the function name accordingly

**Review by passabilities — CHANGES_REQUESTED** [review]

File: contracts/base/escrow/dapps/Aave.sol (hunk 3 above)
Comment (sypeer) [reply contracts/base/escrow/dapps/Aave.sol:86]: Added token balance check to the function

File: contracts/base/escrow/dapps/Aave.sol [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+    /**
+        @notice The deposit, the aToken needs to be approved to have access to the token balance
+        @param tokenAddress address of the token
+        @param amount amount of tokens to deposit
+     */
+    function deposit(address tokenAddress, uint256 amount) public onlyBorrower {
+        IAToken aToken = _getAToken(tokenAddress);
+        IAaveLendingPool aaveLendingPool = _getAavePoolAddress(tokenAddress);
+        uint256 aTokenBalanceBeforeDeposit = aToken.balanceOf(address(this));
+        IERC20(tokenAddress).safeApprove(address(aaveLendingPool), amount);
```
[the remark this answers is not in the document]
Comment (sypeer) [reply contracts/base/escrow/dapps/Aave.sol:50]: Will do this in a separate pr

**Review by passabilities — DISMISSED** [review]

**Review by tpscrpt — APPROVED** [review]
Clean

**Review by passabilities — APPROVED** [review]

**Pull request merged**